### DISCUSSION TITLE
Config option for processes to ignore when confirming close

### DISCUSSION THREAD

**Discussion by lilyball** [ghostty-org/ghostty#2677] [Feature Requests, Ideas]
Ghostty's `confirm-close-surface` skips the close prompt by using shell integration to detect when you're at a prompt. I can see the motivation for this for the login shell, though it also means it'll skip the confirmation for subshells which may or may not be desired. But the problem here is it means it won't skip the prompt for anything that doesn't use the shell integration, either because it's an unsupported shell, or because it's running in a terminal multiplexer like tmux/zellij/screen.

Terminal.app accomplishes this same "don't prompt at the shell" feature through a combination of two things:

1. It skips when you're at the login shell[^1].
2. It has a profile preference in the Shell tab for other processes to ignore, specified by process name. This list defaults to `screen` and `tmux` but is configurable.

That preference for other commands to ignore is the thing that I think would be nice to add to Ghostty (the "login shell" one also seems reasonable too for shells that don't support the shell integration). Having the option to skip terminal multiplexers is nice because closing the window doesn't lose that state, you can just reattach to it in a new window. Being able to include other processes that the user doesn't care about interrupting would be nice too. This does require knowing what processes are controlled by the terminal but I'm assuming that's not hard.

[^1]: I'm not entirely sure how this detection is done. Opening a new window and letting `login` spawn the shell works. Opening a new window using a custom command and specifying the path to my login shell doesn't. Opening a new window using a custom command and specifying `/bin/bash` does. In all 3 cases the shell is the child of the `login` command. It's possible it's consulting `/etc/shells` here for custom commands, since I didn't bother updating that with my shell.

**Comment by mitchellh** [comment]
I think both of these suggestions are good improvements to the confirm close logic. I think both of them also likely involve parsing the process tree, or is there some other way I'm missing? 

If it is parsing the process tree, I suppose the next question is how it's doing this. "How" here is how its getting updates, now how its getting the tree. Since we own the process tree we should be able to easily use syscalls to get child information. But as far as I know, there isn't an evented way to get notification when new child processes start (there is a kqueue event when child processes end), so there must be some kind of polling happening here. 

If it is polling, the question I have is how often and under what conditions...

**Reply by mitchellh** [reply]
Looks like kqueue may support detecting child spawn:

```
EVFILT_PROC    Takes the process ID to monitor as the identifier and the
                    events to watch for in fflags, and returns when the
                    process performs one or more of the requested events.  If
                    a process can normally see another process, it can attach
                    an event to it.  The events to monitor are:

                    NOTE_EXIT
                       The process has exited.

                    NOTE_FORK
                       The process created a child process via fork(2) or sim-ilar similar
                       ilar call.

                    NOTE_EXEC
                       The process executed a new process via [execve(2)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/execve.2.html#//apple_ref/doc/man/2/execve) or
                       similar call.

                    NOTE_SIGNAL
                       The process was sent a signal. Status can be checked
                       via [waitpid(2)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/waitpid.2.html#//apple_ref/doc/man/2/waitpid) or similar call.

                    NOTE_REAP
                       The process was reaped by the parent via [wait(2)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/wait.2.html#//apple_ref/doc/man/2/wait) or
                       similar call.

                    On return, fflags contains the events which triggered the
                    filter.
```

If so, then we'll probably need to update [libxev](https://github.com/mitchellh/libxev) to support this.

**Reply by lilyball** [reply]
fwiw I do think Ghostty should keep track of the process tree anyway, because I really want it to pick up the Terminal.app feature of actually showing me the list of running processes (Terminal.app sticks this in an info window, and lets you copy data from it or send signals to the process group).

I haven't poked at Terminal.app's list in detail, other than to note that it doesn't list children of `tmux`, so what I assume it's doing is keeping track of any process whose controlling terminal is it. If it's using kqueue to track spawns, it must also somehow detect if a child calls `setsid()` to ditch its controlling terminal. I haven't ever looked at terminal devices in great detail so I do wonder if there's not any other way for a terminal to keep track of processes it controls. It's also entirely possible that it simply polls for children.

**Comment by shushtain** [comment]
I would also love for this feature to be implemented. I don't know what Terminal.app is, but there is an option in WezTerm called `skip_close_confirmation_for_processes_named`. It allows to specify shells or apps to ignore.

At least 20 times a day I forget to quit yazi (file manager) before closing the window. Fuming here 🧖‍♂

**Reply by 00-kat** [reply]
(Terminal.app is the default macOS terminal emulator.)

**Reply by shushtain** [reply]
Oh well, then I'm sure it's better to reference something open source for the possible implementations. Besides the WezTerm's config option I mentioned previously, iTerm2 seems to also have it, via the Closing tab in settings. Maybe it would be easier to use something from WezTerm's Rust code, though, rather than Objective-C/Swift, but I'm not sure.

**Reply by rhodes-b** [reply]
Iterm is GPLv2 I would advise referencing it since ghostty is MIT
even with that ghostty is not a rust project so it cannot integrate with rust easily (and there is no reason to have just a tiny bit of rust)

> At least 20 times a day I forget to quit yazi (file manager) before closing the window. Fuming here 🧖‍♂

you can also just disable the popup that happens
https://ghostty.org/docs/config/reference#confirm-close-surface

**Reply by shushtain** [reply]
1) WezTerm is also MIT.

2) Yes, neither is written in Zig, but some languages could be easier to transfer the logic from. I'm not proposing to copy Rust code, that's not what `to reference` means. I'm proposing to _observe_ existing features in other terminals with accessible code, rather than decoding proprietary MacOS implementations. That said, if the feature would even be considered.

3) As I mentioned, I appreciate the safety of confirmations, except for some programs where I don't need it. The same way my browser or an image editor confirms exit only for continuous processes and unsaved files. I thought we are talking about a possible new feature, not helping me read the docs.

4) Yes, I'm creating problems instead of solving them, because unfortunately I can't program in Zig. What I'm asking for is to consider implementing this feature. That's all. Of course, I could make yazi open in WezTerm if I take time to set it up properly, but the feature itself is still quite popular, and that's why many other terminals implement it.

**Reply by rhodes-b** [reply]
And there is no reason to +1 and provide no useful input beyond that wezterm and iterm2 support it (github even lets you add a up arrow which even bumps the thread for the future)
![20250508_23h57m38s_grim](https://github.com/user-attachments/assets/8470d020-bc13-4a9b-a748-719fa9dc6221)

> WezTerm is also MIT.

notice how I left that one out on not having a incompatible license

> Yes, neither is written in Zig, but some languages could be easier to transfer the logic from. I'm not proposing to copy Rust code, that's not what to reference means. I'm proposing to observe existing features in other terminals with accessible code, rather than decoding proprietary MacOS implementations. That said, if the feature would even be considered.

Parsing a process tree isn't some language dependent feature python could do it. Rust and Swift are not vastly different language types compared to zig like functional programming referencing the language makes no sense on which one is easier as such there is no reason to bring them up. Useful info would be wezterm actually does monitor the process tree here is how it does it

> As I mentioned, I appreciate the safety of confirmations, except for some programs where I don't need it. The same way my browser or an image editor confirms exit only for continuous processes and unsaved files. I thought we are talking about a possible new feature, not helping me read the docs.

You said no such thing you just said I would like this feature because I get this popup alot and nowhere in this discussion currently references the ability to just turn off the prompt all together many many people dont read the docs at all I have no reason to assume you have

> Maybe it would be easier to use something from WezTerm's Rust code, though, rather than Objective-C/Swift, but I'm not sure.

generally when I "use" something I'm not going to look how they do it and copy and paste / port that is not what use means

**Comment by shushtain** [comment]
If you don't like my comments, you can just ignore them, or I could be
restricted from posting by someone who has the rights to do so. I'm trying
my best, and the Discussions section, especially the Ideas part, is exactly
the place to post something like this, rather than the Issues section, for
more polished, technical requests.

I **am** an amateur, and I **do** use Python primarily, but I'm still a
valid Ghostty user. I'm "contributing" respectively, not demanding
anything, and adding comments based on what seems useful to me, I can't
predict what would be useful to other people because I don't read minds.

Even if my suggestions are minute, and even if they are stupid, instead of
one "unnecessary" comment that made you so angry we have a whole thread now
with things that don't improve anything. Rules are implemented to make
participating less, nor more toxic. So let's stop.

**Reply by pluiedev** [reply]
Ultimately it's our discussion board and we reserve all rights to minimize comments that only serve as noise. A quick browse at some of the more popular topics like #189 shows you why it's necessary to do so - so that people who are subscribed to an issue are not incessantly pinged or sent an email whenever somebody comments something irrelevant or "I also need this, please magically do what we want." 

This is a discussion board for gathering information, not a chatroom, nor a unidirectional channel for pressure. Ghostty is a volunteer project and that sort of pressure tactic does not magically land wishes in reality, and only wastes people's time. It'll be ready, when it is ready. Nothing personal.
